### PR TITLE
Add no-unreachable-block rule

### LINKED ISSUE #181: Rule: Boolean expressions should not be gratuitous
Implements [RSPEC-2589](https://jira.sonarsource.com/browse/RSPEC-2589)

### PR DESCRIPTION
**Pull request by stas-vilchik**
Fixes #181

### CODE REVIEW

File: sonarts-core/docs/rules/no-unreachable-block.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# no-unreachable-block
+
+Curly braces can be omitted from a one-line block, such as with an `if` statement or `for` loop, but doing so can be misleading and induce bugs.
```
Comment (vilchik-elena) [inline sonarts-core/docs/rules/no-unreachable-block.md]: curly braces?

File: sonarts-core/docs/se/symbolic-execution-design-notes.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# Symbolic Execution Design Notes for SonarTS
+
+* The execution tree is currently visited depth-first and it's not currently using a worklist approach
+
+* Rules are invoked at the end of the full visit of the exploded graph
+** During execution we collect all the program states for each program point
+** After execution the rule (or other callback) is called by providing the program point and the full set of program states
+
+* Literal values are represented as `LiteralSymbolicValue`, which contains the exact value.
+** We assume that once constraints are implemented the `LiteralSymbolicValue` will provide a constraint that corresponds to the value upon request.
```
Comment (vilchik-elena) [inline sonarts-core/docs/se/symbolic-execution-design-notes.md]: @inverno something wrong with formatting. Probably you should use indent, not 2 stars

File: sonarts-core/docs/se/symbolic-execution-design-notes.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+** Other kinds of symbolic values so far are `UndefinedSymbolicValue`...
```
Comment (vilchik-elena) [inline sonarts-core/docs/se/symbolic-execution-design-notes.md]: IMO `Unknown` is better, `Undefined` is javascript value

Comment (vilchik-elena) [reply sonarts-core/docs/se/symbolic-execution-design-notes.md]: Btw why not having just generic `SymbolicValue`?

File: sonarts-core/docs/se/symbolic-execution-design-notes.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+* The SE engine is invoked by providing a list of statements and an initial program state, which, in the case of functions/methods, needs to include the parameters. The responsibility to initialize that state is currently outside the engine
+
+* We are keeping the 'state transition' logic outside of the main graph visit, the expression stack and the program state.
+```
+GraphVisit(OldProgramState) --> ProgramPoint
```
Comment (vilchik-elena) [inline sonarts-core/docs/se/symbolic-execution-design-notes.md]: what is that 3 lines?

File: sonarts-core/src/se/SymbolicExecution.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      callbacks.forEach(callback => callback(programPoint, programStates));
+    });
+  };
+
+  /** 
+   * Visit a program point with a given program state to produce next program state.
+   * @returns `undefined` if the program point with next program state was already visited, 
+   * @returns next program state otherwise
+   */
+  private visitProgramPoint(programPoint: ts.Node, programState: ProgramState) {
```
Comment (SonarTech) [inline sonarts-core/src/se/SymbolicExecution.ts]: MAJOR Refactor this function to use "return" consistently [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS3801)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonarts-core/src/se/programStates.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  public static empty() {
+    return new ProgramState(new Map(), []);
+  }
+
+  private constructor(symbolicValues: SymbolicValues, expressionStack: ExpressionStack) {
+    this.symbolicValues = symbolicValues;
+    this.expressionStack = expressionStack;
+  }
+
+  sv(symbol: ts.Symbol): SymbolicValue | undefined {
```
Comment (vilchik-elena) [inline sonarts-core/src/se/programStates.ts]: can we extend this name a bit?)

File: sonarts-core/src/se/stateTransitions.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  if (expression.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
+    let [value, nextState] = state.popSV();
+    const { getSymbolAtLocation } = program.getTypeChecker();
+    const variable = getSymbolAtLocation(expression.left as ts.Identifier);
+    if (!variable) {
+      return nextState;
+    }
+
+    if (!value) {
+      throw "Assignment without value";
```
Comment (SonarTech) [inline sonarts-core/src/se/stateTransitions.ts]: MAJOR Throwing plain strings (not instances of Error) gives no stack traces [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS3696)

File: sonarts-core/src/se/stateTransitions.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    const variable = getSymbolAtLocation(declaration.name);    
+    if (!variable) {
+      return nextState;
+    }
+
+    if (!value) {
+      value = createUndefinedSymbolicValue();
+    }
+    if (!nextState.hasEmptyStack()) {
+      throw "Non empty stack at the end of declaration " + declaration.getText();
```
Comment (SonarTech) [inline sonarts-core/src/se/stateTransitions.ts]: MAJOR Throwing plain strings (not instances of Error) gives no stack traces [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS3696)

File: sonarts-core/src/se/programStates.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    return new ProgramState(this.symbolicValues, newExpressionStack);
+  }
+
+  popSV(): [SymbolicValue | undefined, ProgramState] {
+    const newExpressionStack = [...this.expressionStack];
+    return [newExpressionStack.pop(), new ProgramState(this.symbolicValues, newExpressionStack)];
+  }
+
+  hasEmptyStack(): boolean {
+    return this.expressionStack.length == 0;
```
Comment (SonarTech) [inline sonarts-core/src/se/programStates.ts]: MAJOR == should be === [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1440)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 4 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonarts-core/src/se/stateTransitions.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as ts from "typescript";
+import * as tsutils from "tsutils";
+import { createLiteralSymbolicValue, createUnknownSymbolicValue, createUndefinedSymbolicValue, createObjectLiteralSymbolicValue } from "./symbolicValues";
+import { ProgramState } from "./programStates";
+
+export function applyExecutors(programPoint: ts.Node, state: ProgramState, program: ts.Program): ProgramState {
```
Comment (SonarTech) [inline sonarts-core/src/se/stateTransitions.ts]: CRITICAL The function applyExecutors has a cyclomatic complexity of 11 which is higher than the threshold of 10 [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1541)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonarts-core/src/se/stateTransitions.ts [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+import * as tsutils from "tsutils";
+import {
+  createLiteralSymbolicValue,
+  createUnknownSymbolicValue,
+  createUndefinedSymbolicValue,
+  createObjectLiteralSymbolicValue,
+} from "./symbolicValues";
+import { ProgramState } from "./programStates";
+
+export function applyExecutors(programPoint: ts.Node, state: ProgramState, program: ts.Program): ProgramState {
```
Comment (SonarTech) [inline sonarts-core/src/se/stateTransitions.ts]: CRITICAL The function applyExecutors has a cyclomatic complexity of 11 which is higher than the threshold of 10 [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1541)

File: sonarts-core/src/se/stateTransitions.ts [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+  createLiteralSymbolicValue,
+  createUnknownSymbolicValue,
+  createUndefinedSymbolicValue,
+  createObjectLiteralSymbolicValue,
+} from "./symbolicValues";
+import { ProgramState } from "./programStates";
+import { PostfixUnaryExpression } from "typescript";
+import { isIdentifier } from "tsutils";
+
+export function applyExecutors(programPoint: ts.Node, state: ProgramState, program: ts.Program): ProgramState {
```
Comment (SonarTech) [inline sonarts-core/src/se/stateTransitions.ts]: CRITICAL The function applyExecutors has a cyclomatic complexity of 12 which is higher than the threshold of 10 [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1541)

File: sonarts-core/src/se/programStates.ts [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+  addFalsyConstraint() {
+    return this.addConstraint(getFalsyConstraint());
+  }
+
+  getConstraints(sv: SymbolicValue) {
+    return this.constraints.get(sv) || [];
+  }
+
+  hasEmptyStack(): boolean {
+    return this.expressionStack.length == 0;
```
Comment (SonarTech) [inline sonarts-core/src/se/programStates.ts]: MAJOR == should be === [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1440)

File: sonarts-core/src/se/programStates.ts [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+  canBeConstrainedTo(constraint: Constraint) {
+    const sv = this.expressionStack[this.expressionStack.length - 1];
+    const constraints = this.getConstraints(sv);
+    return isTruthyConstraint(constraint)
+      ? constraints.every(c => !isFalsyConstraint(c))
+      : constraints.every(c => !isTruthyConstraint(c));
+  }
+
+  hasEmptyStack(): boolean {
+    return this.expressionStack.length == 0;
```
Comment (SonarTech) [inline sonarts-core/src/se/programStates.ts]: MAJOR == should be === [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1440)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonarts-core/src/se/SymbolicExecution.ts [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+  let visits = 0;
+  visitBlock(cfg.start, initialState);
+  if (visitsLimitBreached()) {
+    // Analysis incomplete, it's safer to not report partial results
+    return;
+  } else {
+    return { programNodes, branchingProgramNodes, visits };
+  }
+
+  function visitBlock(block: CfgBlock, programState: ProgramState) {
```
Comment (SonarTech) [inline sonarts-core/src/se/SymbolicExecution.ts]: CRITICAL The function visitBlock has a cyclomatic complexity of 15 which is higher than the threshold of 10 [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1541)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by stas-vilchik** [comment]
Closed in favor of #319

**Pull request closed without merging**